### ISSUE TITLE
Clear console on every build

### ISSUE THREAD

**Issue by dannypike** [openscad/openscad#3905]
It can be very hard to find the beginning of the console output for a build, when it is continually rebuilding because I am editing the source file. Clearing the console manually and then hitting F5 again is one workaround but it is a bit cumbersome and tiresome to do when the rebuild takes a long time.

I would find it very helpful to have an option to auto-clear the console every time that it loads the design. Would that be hard to implement?

**Comment by t-paul** [comment]
Hard to implement, no, but having lots of settings does increase the maintenance effort. So it would be useful to hear some more details of what information is needed from the console output. Maybe some other solution could improve things even more (e.g. like the new ErrorLog window collecting error and warning message of the last run only).

**Comment by MichaelAtOz** [comment]
It used to clear the console all the time, that was also painful, not being able to compare output from a previous compile.

@dannypike reading between your lines, are you using Design/Auto-reload&preview?

> auto-clear the console every time that it loads the design

Is that specifically 'load' ie Reload (auto or manual) and/or Open-file, or more generic ie each F5?

Help to find the start, at top of your file:
```
echo("================================================");
echo(str("==",version(),"==") );
echo("================================================");
```

**Comment by dannypike** [comment]
@t-paul I agree that we need to take care to avoid unnecessary feature-creep because of the potential maintenance effort, and I can see that it may also be nice to be able to compare output from a previous build (so I suggested a setting). The Error Log window is useful for seeing errors but I don't have a problem seeing those in the Console. I tend to build as I work, so I never have lots of errors at any one time and they already stand out in the Console, having a red background.

What I want to be able to have is lots of is echo() calls permanently in the model and to be able to locate them easily, when there are hundreds of lines to scroll through.

I have tried padding my echo() statements as per @MichaelAtOz 's suggestion but I find that gets ugly very quickly when you have lots of them and just creates a tangled mess in the output that is even harder to sort through.

How about changing echo() to:

(a) remove the compulsory "ECHO:" from the start of the line (I can add that, if I want to);
(b) remove the double quotes from the output, when I supply a string (why does it do that?); and
(c) support an empty string which outputs a blank line (that would mean I don't need it to clear the console at all)?

Would that be a solution?

**Comment by MichaelAtOz** [comment]
A lot of people already process the existing echo() output, so that should not change drastically.

There have been requests for an alternate to echo tho.

Option, add a string escape "\c" (?) to clear the console, like the old vt52 commands. ?? (yes I'm old...)

 

  _____

**Comment by dannypike** [comment]
ANSI sequences would be one option (I'm also old enough to have used a VT100 attached to a PDP 11/70), but it'll nasty to implement them all. The CSI stuff is esoteric so it'll probably spawn a whole load of support questions and criticism. It would probably require the addition of a ```$csi``` global value, as well as true string concenation, so there's no saving in maintenance effort by doing this instead of having a new setting.

Even if that is the chosen option, I would still want OpenSCAD to refrain from outputting anything other than what I supply in the parameters, i.e. no "ECHO:" header, and definitely no double quotes unless I specify them. I don't want to prefix every line with an escape sequence that homes the cursor, and erases the line at the start and adds a backspace and clear-to-end-of-line at the end.  Again, a new setting sounds much simpler to me.

In summary, if people want to have the original output because they are processing it, then they switch off the new setting that I propose. The same model file will work in both environments, there's no need to update the parser, implement and maintain a new command, or document and answer questions about weird control sequences.

**Comment by t-paul** [comment]
I'd suggest to keep this issue for the original request and have further discussions about formatting in a separate issue.

**Comment by jordanbrown0** [comment]
Output that omits the "ECHO:" prefix and quotes is covered under #3895.

**Comment by dannypike** [comment]
@jordanbrown0; to me ticket 3895 seems to be about adding an automatic log-to-file option, rather than how to control the text that gets logged.

However, I agree with @t-paul's comment that we shouldn't discuss more than a single issue in each ticket. If there was a way to control the echo() output completely, i.e. without all the decoration. If I could do that, I wouldn't need to have the Clear Console option, because I could insert blank lines to achieve an adequate border between rebuild logs.

So, I'll close this ticket, and open a new one that is specific to formatting.
